### PR TITLE
feat(new source): New fluent source

### LINKED ISSUE #419: New `fluent` source
This will make it easier for users to transition in production environments. We basically want to support the `forward` input and output for Fluentbit. I believe it's the same protocol used for FluentD, but we should investigate further.  They have [a section](https://docs.fluentbit.io/manual/output/forward#fluentd) in their docs that covers this.

- [ ] Add a Fluent transitioning guide that walks through how to do this.

### PR DESCRIPTION
**Pull request by jszwedko**
New source for consuming data from fluentd and fluent-bit via their native [forward protocol](https://github.com/fluent/fluentd/wiki/Forward-Protocol-Specification-v1).

This implementation supports the various formats that the specification allows for but does not implement the authentication protocol (#7532) or the acknowledgment (#7533) parts of the protocol. I left these for follow-up based on user demand given it is a significant divergence from the existing TCP source which just consumes data rather than request/response.

Aside from the source implementation, the other significant change here is extraction of some bits of the `docker_logs` source into a shared `docker` module to use it in the `fluent` source integration tests. The integration tests are more similar to the `docker_logs` integration tests than the other component integration tests in that they each run a separate docker container rather than just standing up one long running one before running the tests as these tests need to run fluentd and fluent-bit with different configurations.

Apologies, I didn't break up the early commits into logical changes, but hopefully the PR itself is reviewable.

Closes #419

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -343,6 +349,8 @@ wasm = ["lucet-runtime", "lucet-wasi", "lucetc", "vector-wasm"]
 # transforms and sinks should depend on this feature.
 kubernetes = ["k8s-openapi", "evmap"]
 
+docker = ["bollard", "dirs-next"]
```
Comment (jszwedko) [inline Cargo.toml:352]: New feature to extract some of the bits from the `docker_logs` source to share with the `fluent` integration tests.

File: src/docker.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
+use bollard::{errors::Error as DockerError, Docker, API_DEFAULT_VERSION};
```
Comment (jszwedko) [inline src/docker.rs:1]: This file contains bits extracted from `docker_logs`.

File: src/sources/docker_logs.rs [hunk 3]
```diff
@@ -1,6 +1,7 @@
 use super::util::MultilineConfig;
 use crate::{
     config::{log_schema, DataType, SourceConfig, SourceContext, SourceDescription},
+    docker::{docker, DockerTlsConfig},
```
Comment (jszwedko) [inline src/sources/docker_logs.rs:4]: Changes to this file are just removals from moving bits into `src/docker.rs`.

File: src/sources/util/tcp.rs [hunk 4]
```diff
@@ -53,17 +53,17 @@ async fn make_listener(
     }
 }
 pub trait IsErrorFatal {
-    fn is_error_fatal() -> bool;
+    fn is_error_fatal(&self) -> bool;
```
Comment (jszwedko) [inline src/sources/util/tcp.rs:56]: Started injecting `self` here as the `fluent` source decides which errors are fatal based on whether they are i/o or decode errors (only i/o errors are fatal).

File: tests/config.rs [hunk 5]
```diff
@@ -760,11 +760,7 @@ async fn parses_sink_full_es_basic_auth() {
     .unwrap();
 }
 
-#[cfg(all(
-    feature = "docker",
```
Comment (jszwedko) [inline tests/config.rs:764]: Not sure why this feature constraint was here as the feature didn't exist, but it flagged this test as failing to compile when I reintroduced it.

**Review by spencergilbert — APPROVED** [review]
Walked through this with @jszwedko and seems good, but definitely should have someone more rust savvy give it a once-over :)

File: src/sources/fluent.rs [hunk 6]
```diff
@@ -0,0 +1,1040 @@
...
+
+    fn decoder(&self) -> Self::Decoder {
+        FluentDecoder::new()
+    }
+
+    fn build_event(&self, frame: FluentFrame, host: Bytes) -> Option<Event> {
+        let mut log = LogEvent::from(frame);
+        if log.get(log_schema().host_key()).is_none() {
+            log.insert(log_schema().host_key(), host);
+        }
```
Comment (bruceg) [inline src/sources/fluent.rs:94]: Is this better with the `entry` interface?
```suggestion
        log.entry(log_schema().host_key()).or_insert(host);
```

File: src/sources/fluent.rs [hunk 7]
```diff
@@ -0,0 +1,1040 @@
...
+}
+
+impl Decoder for FluentDecoder {
+    type Item = FluentFrame;
+    type Error = DecodeError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        if !self.unread_frames.is_empty() {
+            return Ok(self.unread_frames.pop_front());
+        }
```
Comment (bruceg) [inline src/sources/fluent.rs:233]: Don't need to check `is_empty` first:
```suggestion
        if let Some(frame) = self.unread_frames.pop_front() {
            return Ok(frame);
        }
```

**Review by bruceg — APPROVED** [review]
Nothing obviously bad jumped out at me, just a couple of minor tweaks. The real test is integrating with fluentd and fluent-bit, which you are testing.

Can fluentd run in podman as well as docker? If so, is there any option to allow for testing under podman?

File: src/sources/fluent.rs (hunk 6 above)
Comment (jszwedko) [reply src/sources/fluent.rs:94]: 👍 I always forget about that interface.

**Comment by jszwedko** [comment]
> Can fluentd run in podman as well as docker? If so, is there any option to allow for testing under podman?

I took a brief look at this and it seems like it could possibly work but I wasn't able to get podman running on my Linux system (I ran into this issue: https://stackoverflow.com/questions/65775075/problems-running-podman-in-ubuntu-20-04 but wasn't able to resolve it).

I was able to run `podman system service -t 0  unix:///tmp/podman.sock` to run a docker compatible API and then `export DOCKER_HOST to point to this though and things appeared to be working (minus the issue I ran into with the uid/gids). I do think we'll need this unreleased change in podman: https://github.com/containers/podman/pull/9972 to configure `host.docker.internal` to point to the host system (needed since we can't use 127.0.0.1 to indicate vector on OSX given that docker is running in a VM).

Would you be able to try it out?

**Comment by jszwedko** [comment]
Kicked off integration test to make sure it passes in CI: https://github.com/timberio/vector/actions/runs/886440980

**Pull request merged**